extmod/moddeflate.c

```diff
@@ -0,0 +1,411 @@
+    return c;
+}
+
+STATIC bool deflateio_init_read(mp_obj_deflateio_t *self) {
+    if (self->read) {
+        return true;
+    }
+
+    mp_get_stream_raise(self->stream, MP_STREAM_OP_READ);
+
+    self->read = m_new_obj(mp_obj_deflateio_read_t);
+    memset(&self->read->decomp, 0, sizeof(self->read->decomp));
+    self->read->decomp.source_read_data = self;
+    self->read->decomp.source_read_cb = deflateio_read_stream;
+    self->read->eof = false;
+
+    // Don't modify self->window_bits as it may also be used for write.
+    int wbits = self->window_bits;
+
+    // Parse the header if we're in NONE/ZLIB/GZIP modes.
+    if (self->format != DEFLATEIO_FORMAT_RAW) {
+        int header_wbits = wbits;
+        int header_type = uzlib_parse_zlib_gzip_header(&self->read->decomp, &header_wbits);
+        if ((self->format == DEFLATEIO_FORMAT_ZLIB && header_type != UZLIB_HEADER_ZLIB) || (self->format == DEFLATEIO_FORMAT_GZIP && header_type != UZLIB_HEADER_GZIP)) {
+            return false;
+        }
+        if (wbits == 0 && header_wbits < 15) {
+            // If the header specified something lower than the default, then
+            // use that instead.
+            wbits = header_wbits;
+        }
+    }
+
+    if (wbits == 0) {
+        wbits = DEFLATEIO_DEFAULT_WBITS;
+    }
+
+    size_t window_len = 1 << wbits;
+    self->read->window = m_new(uint8_t, window_len);
+
+    uzlib_uncompress_init(&self->read->decomp, self->read->window, window_len);
+
+    return true;
+}
+
+#if MICROPY_PY_DEFLATE_COMPRESS
+STATIC void deflateio_out_byte(void *data, uint8_t b) {
+    mp_obj_deflateio_t *self = data;
+    const mp_stream_p_t *stream = mp_get_stream(self->stream);
+    int err;
+    mp_uint_t ret = stream->write(self->stream, &b, 1, &err);
+    if (ret == MP_STREAM_ERROR) {
+        mp_raise_OSError(err);
+    }
+}
+
+STATIC bool deflateio_init_write(mp_obj_deflateio_t *self) {
+    if (self->write) {
+        return true;
+    }
+
+    const mp_stream_p_t *stream = mp_get_stream_raise(self->stream, MP_STREAM_OP_WRITE);
+
+    self->write = m_new_obj(mp_obj_deflateio_write_t);
+    self->write->input_len = 0;
+
+    int wbits = self->window_bits;
+    if (wbits == 0) {
+        wbits = DEFLATEIO_DEFAULT_WBITS;
+    }
+    size_t window_len = 1 << wbits;
+    self->write->window = m_new(uint8_t, window_len);
+
+    uzlib_lz77_init(&self->write->lz77, self->write->window, window_len);
+    self->write->lz77.dest_write_data = self;
+    self->write->lz77.dest_write_cb = deflateio_out_byte;
+
+    // Write header if needed.
+    mp_uint_t ret = 0;
+    int err;
+    if (self->format == DEFLATEIO_FORMAT_ZLIB) {
+        // -----CMF------  ----------FLG---------------
+        // CINFO(5) CM(3)  FLEVEL(2) FDICT(1) FCHECK(5)
+        uint8_t buf[] = { 0x08, 0x80 };     // CM=2 (deflate), FLEVEL=2 (default), FDICT=0 (no dictionary)
+        buf[0] |= MAX(self->window_bits - 8, 1) << 4;     // base-2 logarithm of the LZ77 window size, minus eight.
```

Comment:
Should this use `wbits`?  In case `self->window_bits==0`?

Reply:
Good catch. Fixed.